src/gs-discovery-feed-content-provider.c

```diff
@@ -0,0 +1,295 @@
+
+static guint
+get_discovery_feed_app_kudo_score (GsApp *app)
+{
+	guint i = 0;
+	guint score = 0;
+	guint64 kudos = gs_app_get_kudos (app);
+
+	for (; i < discovery_feed_app_kudos_len; ++i) {
+		if ((kudos & discovery_feed_app_kudos[i].kudo) != 0) {
+			score += discovery_feed_app_kudos[i].weight;
+		}
+	}
+
+	return score;
+}
+
+static gint
+search_sort_by_kudo_cb (GsApp *app1, GsApp *app2, gpointer user_data)
+{
+	guint pa, pb;
+	pa = get_discovery_feed_app_kudo_score (app1);
+	pb = get_discovery_feed_app_kudo_score (app2);
+	if (pa < pb)
+		return 1;
+	else if (pa > pb)
+		return -1;
+	return 0;
+}
+
+static gchar *
+get_app_thumbnail_cached_filename (GsApp *app)
+{
+	const gchar *popular_background_url = gs_app_get_metadata_item (app, "GnomeSoftware::popular-background");
+	g_autofree char *tile_cache_hash = NULL;
+	g_autofree char *cache_identifier = NULL;
+	g_autofree char *url_basename = NULL;
+	char *cache_filename = NULL;
+
+	url_basename = g_path_get_basename (popular_background_url);
+	tile_cache_hash = g_compute_checksum_for_string (G_CHECKSUM_SHA256,
+	                                                 popular_background_url,
+	                                                 -1);
+	cache_identifier = g_strdup_printf("%s-%s", tile_cache_hash, url_basename);
+	cache_filename = gs_utils_get_cache_filename ("eos-popular-app-thumbnails",
+	                                              cache_identifier,
+	                                              GS_UTILS_CACHE_FLAG_NONE,
+	                                              NULL);
+
+	/* Check to see if the file exists in the cache at the time we called this
+	 * function. If it does, then we can use this app in the results
+	 * since it will have a thumbnail */
+	if (g_file_test (cache_filename, G_FILE_TEST_EXISTS)) {
+		g_debug("Hit cache for thumbnail when loading discovery feed %s: %s", popular_background_url, cache_filename);
+		return cache_filename;
+	}
+
+    return cache_filename;
+}
+
+static const guint N_APPS_TO_INCLUDE_IN_RESULT = 5;
+
+static void
+search_done_cb (GObject *source,
+		GAsyncResult *res,
+		gpointer user_data)
+{
+	PendingSearch *search = user_data;
+	GsDiscoveryFeedContentProvider *self = search->provider;
+	guint i;
+	GVariantBuilder builder;
+	guint app_list_length;
+	guint count;
+	g_autoptr(GsAppList) list = NULL;
+
+	list = gs_plugin_loader_search_finish (self->plugin_loader, res, NULL);
+	if (list == NULL) {
+		g_dbus_method_invocation_return_value (search->invocation, g_variant_new ("(as)", NULL));
+		pending_search_free (search);
+		g_application_release (g_application_get_default ());
+		return;
+	}
+
+	app_list_length = gs_app_list_length (list);
+	count = 0;
+
+	/* sort by kudos, as there is no ratings data by default */
+	gs_app_list_sort (list, search_sort_by_kudo_cb, NULL);
+
+	g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{sv}"));
+	for (i = 0; i < app_list_length; i++) {
+	        gchar *app_thumbnail_filename;
```

Comment:
I'm not sure if we can actually - it is a loop variable and I'm pretty sure that `g_autofree` only happens on function exit and not scope exit.